File: applications/holoscan_ros2/adi/a2baudio/cpp/a2b_audio_ros2_subscriber.cpp
#include <algorithm>
#include <cstddef>

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/float32_multi_array.hpp>

class A2BAudioRos2Subscriber : public rclcpp::Node {
public:
  A2BAudioRos2Subscriber() : Node("a2b_audio_ros2_subscriber") {
    raw_subscription_ = create_subscription<std_msgs::msg::Float32MultiArray>(
        "a2b_audio/raw", 10, [this](const std_msgs::msg::Float32MultiArray::ConstSharedPtr message) {
          RCLCPP_INFO(get_logger(), "raw audio: %zu samples", message->data.size());
        });
    beam_subscription_ = create_subscription<std_msgs::msg::Float32MultiArray>(
        "a2b_audio/beamformed", 10,
        [this](const std_msgs::msg::Float32MultiArray::ConstSharedPtr message) {
          RCLCPP_INFO(get_logger(), "beamformed audio: %zu samples", message->data.size());
        });
  }

private:
  rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr raw_subscription_;
  rclcpp::Subscription<std_msgs::msg::Float32MultiArray>::SharedPtr beam_subscription_;
};

int main(int argc, char** argv) {
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<A2BAudioRos2Subscriber>());
  rclcpp::shutdown();
  return 0;
}

File: applications/holoscan_ros2/adi/a2baudio/python/a2b_audio_ros2_subscriber.py
import rclpy
from rclpy.node import Node
from std_msgs.msg import Float32MultiArray


class A2BAudioRos2Subscriber(Node):
    def __init__(self):
        super().__init__("a2b_audio_ros2_subscriber")
        self.create_subscription(Float32MultiArray, "a2b_audio/raw", self.on_raw, 10)
        self.create_subscription(Float32MultiArray, "a2b_audio/beamformed", self.on_beamformed, 10)

    def on_raw(self, message):
        self.get_logger().info("raw audio: %d samples" % len(message.data))

    def on_beamformed(self, message):
        self.get_logger().info("beamformed audio: %d samples" % len(message.data))


def main():
    rclpy.init()
    node = A2BAudioRos2Subscriber()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
